### DISCUSSION TITLE
Async SQLAlchemy engine doesn't seem to close connections fully when dispose is called using, mysql and aiomysql

### DISCUSSION THREAD

**Discussion by mdotdi** [sqlalchemy/sqlalchemy#10457] [Usage Questions]
Hi

Whilst using SQLAlchemy async with mysql and aiomysql I noticed the following warning sin the Mariadb logs when dispose was called on my tests:
`2023-10-11 18:46:52 16 [Warning] Aborted connection 16 to db: 'test-db' user: 'test-user' host: '172.17.0.1' (Got an error reading communication packets)`

I took the sample code from https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html and made a few changes to work with MariaDB/MySQL and again saw the warning when engine.dispose is called, however this can be validated with a shortened version of the code.

```
import asyncio

from sqlalchemy.ext.asyncio import AsyncAttrs
from sqlalchemy.ext.asyncio import async_sessionmaker
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column


class Base(AsyncAttrs, DeclarativeBase):
    pass


class TestTable(Base):
    __tablename__ = "test_table"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]


async def async_main() -> None:
    engine = create_async_engine(
        "mysql+aiomysql://test-user:test-password@127.0.0.1/test-db",
        echo=True,
    )

    # async_sessionmaker: a factory for new AsyncSession objects.
    # expire_on_commit - don't expire objects after transaction commit
    async_session = async_sessionmaker(engine, expire_on_commit=False)

    async with engine.begin() as conn:
        pass


    # for AsyncEngine created in function scope, close and
    # clean-up pooled connections
    await engine.dispose()


asyncio.run(async_main())
```

I am using python 3.10 python package versions I had installed:

- aiomysql==0.2.0
- greenlet==3.0.0
- mysql==0.0.3
- mysqlclient==2.2.0
- PyMySQL==1.1.0
- SQLAlchemy==2.0.21
- typing_extensions==4.8.0

I'm not sure if this is expected

Thanks in advance

**Comment by CaselIT** [comment]
Hi,

Can you reproduce also with the other async driver asyncmy?

**Reply by mdotdi** [reply]
Hi
Yes I get the same warning when using asyncmy
Thanks
Matt

**Reply by CaselIT** [reply]
And this happens only with async drtivers? If you use a sync one you don't get that warning in the db?

**Reply by mdotdi** [reply]
Yes that's correct with sync the warning isn't seen

```
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column


class Base(DeclarativeBase):
    pass


class TestTable(Base):
    __tablename__ = "test_table"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]


def main() -> None:
    engine = create_engine(
        "mysql+pymysql://test-user:test-password@127.0.0.1/test-db",
        echo=True,
    )

    # async_sessionmaker: a factory for new AsyncSession objects.
    # expire_on_commit - don't expire objects after transaction commit
    async_session = sessionmaker(engine, expire_on_commit=False)

    with engine.begin() as conn:
        pass


    # for AsyncEngine created in function scope, close and
    # clean-up pooled connections
    engine.dispose()

main()
```

**Reply by CaselIT** [reply]
@zzzeek ideas here?

**Reply by zzzeek** [reply]
step one is make a test case using the async driver alone, call close(), see what MySQL reports

**Comment by mastamatto** [comment]
Any updates on this? I'm running into this same issue in version 2.0.23. Are there any workarounds to this? This creates hundreds of 'aborted connections' when using the 'engine.dispose' method.

**Reply by zzzeek** [reply]
dispose wont clean up connections that arent checked into the pool, so without a runnable example we can't say much on what is actually going on

**Reply by mastamatto** [reply]
Here is a quick example:
```python
import asyncio
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy import text
from sqlalchemy.pool import AsyncAdaptedQueuePool

def make_engine(url):
    engine = create_async_engine(
        url             = url,
        echo            = True,
        poolclass       = AsyncAdaptedQueuePool,
        max_overflow    = 0,
        pool_size       = 1,
        pool_recycle    = -1,
        pool_timeout    = 30,
        connect_args    = {'connect_timeout': 240}
    )
    return engine

async def execute(engine, query):
    data = []
    async with engine.connect() as connection:
        result = await connection.execute(query)
        await connection.commit()
        if result.returns_rows:
            keys    = list(result.keys())
            raw     = result.fetchall()
            data    = [dict(zip(keys, x)) for x in raw]
    return data

async def run(query, url):
    query       = text(query)
    engine      = make_engine(url)
    response    = None
    try:
        response = await execute(engine, query)
    except Exception as e:
        return e.__repr__()
    finally:
        await engine.dispose(close=True)
    return response

async def main():
    query = "show tables"
    url = "some url"
    response = await run(query, url)
    
asyncio.run(main())
```
Sqlalchemy version is 2.0.23 (tested on .25 as well). MariaDB version 10.6.12.

**Reply by gordthompson** [reply]
Hi @mastamatto . I don't get any warnings when I run your code.

Python 3.11
SQLAlchemy 2.0.25
aiomysql 0.2.0

hitting

10.5.5-MariaDB-1:10.5.5+maria~focal

**Reply by mastamatto** [reply]
The warnings will pop up on the mariadb side if you check the server logs

**Reply by gordthompson** [reply]
Ah, okay, I found it in /var/log/syslog

> Jan 13 15:10:46 xubu-22-04 mariadbd[855]: 2024-01-13 15:10:46 35 [Warning] Aborted connection 35 to db: 'test' user: 'scott' host: '192.168.0.179' (Got an error reading communication packets)

**Reply by gordthompson** [reply]
FWIW, I get the same server-side warning when I use the [Basic Example](https://github.com/aio-libs/aiomysql#basic-example) code from the aiomysql README (no SQLAlchemy involved):

```python
import asyncio
import aiomysql


async def test_example(loop):
    pool = await aiomysql.create_pool(host='192.168.0.195', port=3306,
                                      user='scott', password='tiger',
                                      db='test', loop=loop)
    async with pool.acquire() as conn:
        async with conn.cursor() as cur:
            await cur.execute("SELECT 42;")
            print(cur.description)
            (r,) = await cur.fetchone()
            assert r == 42
    pool.close()
    await pool.wait_closed()


loop = asyncio.get_event_loop()
loop.run_until_complete(test_example(loop))
```

/var/log/syslog contains

> Jan 13 15:25:38 xubu-22-04 mariadbd[855]: 2024-01-13 15:25:38 36 [Warning] Aborted connection 36 to db: 'test' user: 'scott' host: '192.168.0.179' (Got an error reading communication packets)

**Reply by mastamatto** [reply]
Ah, so it's an issue with aiomysql? I found the following issue brought up on aiomysql [here](https://github.com/aio-libs/aiomysql/issues/166). It seems the change is:

Replace `cnx.close()` with `await cnx.ensure_closed()`

I wonder how this would be done via sqlalchemy?

**Reply by gordthompson** [reply]
@mastamatto - Good find! The server-side warning went away after I changed the "Basic Example" above to

```python
import asyncio
import aiomysql


async def test_example(loop):
    conn = await aiomysql.connect(host='192.168.0.195', port=3306,
                                  user='scott', password='tiger',
                                  db='test', loop=loop)
    async with conn.cursor() as cur:
        await cur.execute("SELECT 42;")
        print(cur.description)
        (r,) = await cur.fetchone()
        assert r == 42

    await conn.ensure_closed()


loop = asyncio.get_event_loop()
loop.run_until_complete(test_example(loop))
```

**Reply by gordthompson** [reply]
> I wonder how this would be done via sqlalchemy?

It could be as simple as this:

```patch
Index: lib/sqlalchemy/dialects/mysql/aiomysql.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lib/sqlalchemy/dialects/mysql/aiomysql.py b/lib/sqlalchemy/dialects/mysql/aiomysql.py
--- a/lib/sqlalchemy/dialects/mysql/aiomysql.py	(revision 2c8a05244a8f7919c3954e0b6e1efaa996792772)
+++ b/lib/sqlalchemy/dialects/mysql/aiomysql.py	(date 1705249675220)
@@ -69,8 +69,7 @@
         await_(self._connection.autocommit(value))
 
     def close(self):
-        # it's not awaitable.
-        self._connection.close()
+        await_(self._connection.ensure_closed())
 
 
 class AsyncAdapt_aiomysql_dbapi:
```

It makes the server-side warning go away for your [example code](https://github.com/sqlalchemy/sqlalchemy/discussions/10457#discussioncomment-8119815) and nothing blew up, but async is not really "my thing" so let's wait to see what @zzzeek or @CaselIT might have to say.

**Reply by CaselIT** [reply]
that seems to be what should be called. close is just closing the socket on the client without sending any closure to the server. What a peculiar (to be polite) design..

https://github.com/aio-libs/aiomysql/blob/83aa96e12b1b3f2bd373f60a9c585b6e73f40f52/aiomysql/connection.py#L335C1-L351C21

@gordthompson can you open a PR with that?

**Reply by gordthompson** [reply]
https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5109

**Reply by zzzeek** [reply]
It looks like they are using close to mean "terminate" so we should accommodate both elements , since we have close/terminate as well

**Reply by CaselIT** [reply]
Oh, right. So it should look something like asyncpg

**Reply by zzzeek** [reply]
We should probably generalize this pattern across all the asyncio dialects in 2.1, that is, in terminate, check if we are async, do await close, catch everything for that, fallback to sync terminate, etc.

**Reply by mastamatto** [reply]
I noticed that this same error happens with asyncmy as well. (has same solution fortunately)
change `self._connection.close()` to `await_(self._connection.ensure_closed())`

**Reply by kgadek93** [reply]
Hi, I'm using latest version:
SQLAlchemy                                   2.0.27
and I still see this issue. 

Other packages:
- aiomysql                                     0.2.0
- mysqlclient                                  2.2.4
- PyMySQL                                      1.1.0
- SQLAlchemy                                   2.0.27

**Comment by zzzeek** [comment]
hi -

this fix does not prevent connections from being ungracefully terminated in all cases.   if connections are garbage collected before being explicitly closed within the event loop (by using context managers or judicious use of `await connection.close()` within the event loop), or if the python process terminates ahead of connections being closed, the mysql server will still show these messages.  What the fix does do is that when SQLAlchemy is able to emit a full graceful close within the event loop, it uses the API that the asyncmy/aiomysql drivers provide that will perform a full close conversation.

> Other packages:
>    aiomysql 0.2.0
>   mysqlclient 2.2.4
>    PyMySQL 1.1.0
>    SQLAlchemy 2.0.27

so that's three different mysql drivers and only aiomysql there is actually asyncio.  aiomysql might have pymysql as a dependency.

the fix can be demonstrated with this program, please try it out

```py

import asyncio

from sqlalchemy.ext.asyncio import create_async_engine

async def async_main():
    engine = create_async_engine(
        "mysql+asyncmy://scott:tiger@localhost/test",
        echo=True,
    )

    _ = await asyncio.gather(*[go(engine) for i in range(10)])
    await engine.dispose()

    print("disposed engine")


async def go(engine):
    async with engine.begin() as conn:
        result = await conn.exec_driver_sql("select 1")
        result.all()



asyncio.run(async_main())

```

running it before the fix, mysql's log shows:

```
2024-02-15  8:39:07 29 [Warning] Aborted connection 29 to db: 'test' user: 'scott' host: 'localhost' (Got an error reading communication packets)
2024-02-15  8:39:07 28 [Warning] Aborted connection 28 to db: 'test' user: 'scott' host: 'localhost' (Got an error reading communication packets)
2024-02-15  8:39:07 30 [Warning] Aborted connection 30 to db: 'test' user: 'scott' host: 'localhost' (Got an error reading communication packets)
2024-02-15  8:39:07 31 [Warning] Aborted connection 31 to db: 'test' user: 'scott' host: 'localhost' (Got an error reading communication packets)
2024-02-15  8:39:07 32 [Warning] Aborted connection 32 to db: 'test' user: 'scott' host: 'localhost' (Got an error reading communication packets)
2024-02-15  8:39:07 23 [Warning] Aborted connection 23 to db: 'test' user: 'scott' host: 'localhost' (Got an error reading communication packets)
2024-02-15  8:39:07 24 [Warning] Aborted connection 24 to db: 'test' user: 'scott' host: 'localhost' (Got an error reading communication packets)
2024-02-15  8:39:07 25 [Warning] Aborted connection 25 to db: 'test' user: 'scott' host: 'localhost' (Got an error reading communication packets)
2024-02-15  8:39:07 26 [Warning] Aborted connection 26 to db: 'test' user: 'scott' host: 'localhost' (Got an error reading communication packets)
2024-02-15  8:39:07 27 [Warning] Aborted connection 27 to db: 'test' user: 'scott' host: 'localhost' (Got an error reading communication packets)

```

running it after the fix, mysql's log shows

```

```


hopefully this clarifies
